Cache API

penalosa opened the pull request:
Initial implementation wiring together the Cache API and the `cacheApiOutbound` binding. I've tested this with a WIP Miniflare plugin, and it seems to send the cache API requests out as expected. The approach I've taken may well not be the correct one (I've used a special `subrequestChannel` to open an HTTP client—perhaps there's a different way to separate out cache requests?).

kentonv commented on src/workerd/api/cache.c++:
```diff
@@ -504,13 +513,8 @@ kj::Own<kj::HttpClient> Cache::getHttpClient(IoContext& context,
-    return cacheClient->getDefault(kj::mv(cfBlobJson), span);
-  });
-  httpClient = httpClient.attach(kj::mv(span), kj::mv(cacheClient));
+  auto httpClient = context.getHttpClient(IoContext::CACHE_CHANNEL, true, kj::mv(cfBlobJson), operationName);
```
Sorry, this isn't the right approach. The code here should not have to change. Instead you should be implementing `IoChannelFactory::getCache()` in `server.c++`. The only changes that should be needed in this PR are to `server.c++` and `worker.capnp`.

penalosa replied:
Ah, I see. How should the `getCache` implementation behave? Internally it looks like it's an http client, would that work? Is the approach of defining a special subrequest channel along the right lines?

kentonv commented on src/workerd/server/workerd.capnp:
```diff
@@ -466,7 +469,7 @@ struct Worker {
   }
 
-  durableObjectUniqueKeyModifier @8 :Text;
+  durableObjectUniqueKeyModifier @9 :Text;
```
In Cap'n Proto, the purpose of these `@N` numbers is to maintain binary backwards-compatibility of the serialization as new fields are added. If you change the numbers, you defeat the purpose -- this will break compatibility. Instead, leave the numbers the same on existing fields, and assign the next number in sequence to your new field. The numbers do not have to appear in order in the file -- in fact, that's why we have the numbers, so that you can still list fields in logical order rather than exactly the order in which they were added.

penalosa commented:
Okay—attempt 2 😂. Thanks to @harrishancock for walking me through some of how this should work. I've implemented `getCache()` in `server.c++`, which relies on a `cacheOutbound` being set in the capnp config file. Still to do is to implement `getCache()` as a no-op if no cache service is defined.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1303,10 +1304,79 @@ private:
+      uint spliceCount = 0;
+      auto headersCopy = headers.cloneShallow();
+      KJ_IF_MAYBE(name, cacheName) {
+        spliceParts[spliceCount++] = "{\"namespace\":\""_kj;
```
I know you copied this splice stuff from internal code, but the place you copied from was a lot more complicated, and in this case you could do something much cleaner:

```
kj::String headerValue;
KJ_IF_MAYBE(name, cacheName) {
  headerValue = kj::str("{\"namespace\":\"", *name, "\"}");
} else {
  headerValue = kj::str("{}");
}
```

Better yet, if the only purpose of this header is to encode the namespace name, maybe don't use JSON at all. You could have the header be named `X-Cache-Namespace` and have it just contain the name raw. If no namespace is specified, then leave the header off.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1303,10 +1304,79 @@ private:
   kj::Own<CacheClient> getCache() override {
-    // TODO(beta): Cache API emulation.
-    JSG_FAIL_REQUIRE(Error, "The cache API is not yet implemented.");
+     auto& channels = KJ_REQUIRE_NONNULL(ioChannels.tryGet<LinkedIoChannels>(),
```
Indentation is off.

kentonv commented on src/workerd/server/workerd.capnp:
```diff
@@ -424,6 +424,9 @@ struct Worker {
   # Where should the global "fetch" go to? The default is the service called "internet", which
   # should usually be configured to talk to the public internet.
 
+  cacheOutbound @11 :ServiceDesignator = "cache";
```
```suggestion
  cacheApiOutbound @11 :ServiceDesignator = "cache";
```

Let's be clearer that this is specifically about the Cache API, not any implicit caching.

kentonv replied:
Giving this a default value of `cache` will mean that all existing configs that don't declare any such service will break. I think this should default to null, in which case the cache API is considered disabled.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1303,10 +1304,79 @@ private:
+    }
+
+  private:
+    Service& parent;
```
Why is this called `parent`? Maybe `cacheService` instead?

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1303,10 +1304,79 @@ private:
+        kj::StringPtr cacheName, kj::Maybe<kj::String> cfBlobJson,
+        kj::Maybe<Tracer::Span&> parentSpan) override {
+
+      auto encodedName = kj::encodeUriComponent(cacheName);
```
Can you move this encoding step to happen inside addCacheNameHeader? So it's with the other code related to how this is used.

kentonv commented:
This new logic should be tested in `server-test.c++`.

penalosa commented:
@kentonv I've addressed your comments, and am working on adding testing now.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1304,79 +1304,72 @@ private:
 
   private:
-    Service& parent;
+    Service &cacheService;
```
In C++ it's generally recommended that type modifiers like `*` and `&` are grouped with the type, so this should be `Service& cacheService`, not `Service &cacheService`. (In C++ we say "foo is a pointer to T", whereas in C people were encouraged to think "the thing foo points at is a T", hence in C style you would write `T *foo` but in C++ you write `T* foo`.)

It looks like this commit systematically changed all the types in your commit to use the wrong style. Did you run some sort of auto-formatter?

kentonv commented:
It looks like you ran some sort of auto-formatter as part of your commit addressing the review comments. There are two problems with this:

1. The formatter isn't configured right for our style. (Unfortunately, we don't know of any formatter than gets our style exactly right, and therefore I don't recommend using an auto-formatter at all. But in this case the formatter is getting some basic things wrong.)
2. The formatting changes add a lot of noise to the commit which makes it hard for me to see the important changes. Per the contributing guide, it's really improtant to separate noisy changes into separate commits for reviewability.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1304,79 +1304,72 @@ private:
 
-      return kj::heap<CacheHttpClientImpl>(
-          parent,  kj::mv(encodedName), kj::mv(cfBlobJson), parentSpan);
+      return kj::heap<CacheHttpClientImpl>(cacheService, kj::str(cacheName),
```
Oh, now I see why `cacheName` was being encoded here rather than later... because we have to allocate a copy anyway, so if we can combine the encoding step (which also allocates a copy) then we avoid a redundant copy. Sorry, I think I mislead you by suggesting that logic should have been moved later...

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1304,79 +1304,72 @@ private:
+      KJ_IF_MAYBE (name, cacheName) {
+        auto encodedName = kj::encodeUriComponent(*name);
+
+        headersCopy.set(IoContext::current().getHeaderIds().cfCacheNamespace,
```
I don't think it's necessarily safe to assume that there is a current `IoContext` here. It would be better if the header ID could be passed down to `CacheClientImpl` via its constructor.

kentonv replied:
To be clear, it shouldn't be necessary to use `IoContext::current()` at all... I think somewhere further up the stack you should be able to access the `ThreadContext` directly, or maybe the `HttpHeaderTable::Builder` directly.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1304,79 +1304,72 @@ private:
+    auto &channels = KJ_REQUIRE_NONNULL(ioChannels.tryGet<LinkedIoChannels>(),
+                                        "link() has not been called");
+    auto &cache =
+        JSG_REQUIRE_NONNULL(channels.cache, Error, "No Cache was configured");
```
`channels.cache` is not a `kj::Maybe`, it's a raw pointer. The "assert nonnull" asserts work on pointers but it's good style to always use a `Maybe` when nullness is possible.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -1861,14 +1854,19 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
 
-    Service& cacheApi = lookupService(conf.getCacheOutbound(),
-        kj::str("Worker \"", name, "\"'s cacheOutbound"));
+    if (conf.getCacheApiOutbound().getName() != nullptr) {
```
Please use `conf.hasCacheApiOutbound()` to check if the field is present.

kentonv commented:
Still needs tests.

penalosa commented:
@kentonv I've addressed those comments, and added some basic tests. I had to slightly modify the test harness to get it to run on macOS (change `kj_test` to use `linkstatic = 1`), but I'll split that out into a separate PR (it's not related to this, and there might be a better way to make dynamic linking work on macOS)

kentonv commented on src/workerd/server/server-test.c++:
```diff
@@ -2103,6 +2103,98 @@ KJ_TEST("Server: disk service allow dotfiles") {
+    subreq.send(R"(
+      HTTP/1.1 200 OK
+      CF-Cache-Status: HIT
+      Content-Size: 6
```
```suggestion
      Content-Length: 6
```

This is why `Transfer-Encoding: chunked` shows up below...

kentonv commented on src/workerd/server/server-test.c++:
```diff
@@ -2103,6 +2103,98 @@ KJ_TEST("Server: disk service allow dotfiles") {
+
+  )"_blockquote);
+
+}
```
Let's make sure to test `caches.open()` to make sure the name header comes through as expected.

penalosa commented:
I can build the tests fine, but they fail at runtime because they're not able to find symbols from lol html

mhart commented:
Tests haven't been working on Mac – at least in a while. See https://github.com/cloudflare/workerd/issues/107

jclee commented:
Per a separate chat, I have to wonder if maybe the `rust-deps` dependency in `rust-deps/BUILD.bazel` needs to be `rust_library` instead of `rust_static_library`, for non-statically built executables to dynamically link correctly when run?  I don't have a mac dev environment to test this in, though.

I assume the `workerd` executable itself works correctly on macOS, though, and I'm not sure why that would be, unless `bazel build` is somehow defaulting to static builds and `bazel test` is defaulting to dynamic builds?

penalosa commented:
I've done a bit more digging, and it looks like `bazel build` defaults to static, while `bazel test` defaults to dynamic (for given values of static and dynamic-`bazel test` is definitely _more_ dynamic, but I recall from a separate conversation that `bazel build` isn't fully static either).

(As an aside, it's quite possible that I'm wrong, and completely misinterpreting the output of `otool -L`-do please correct me if that's the case)

In terms of whether the `rust_static_library` rule is the culprit; changing to `rust_library` fixes the original issue around not finding symbols relating to lol html. However, it's then unable to find the symbols for rust alloc, and so I wonder if there's something else in the rust bazel setup that needs a similar change.

penalosa commented:
@kentonv I think that should be all the PR comments addressed now
